DOC-10466 - Add Information on Minimum Memory Resident Ratio (#2925)
couchbase/docs-server: modules/learn/pages/buckets-memory-and-storage/memory.adoc

[#service-memory-quotas]
== Service Memory Quotas

> You must configure your memory quota allocations for each service in Couchbase Server, except for the xref:services-and-indexes/services/query-service.adoc[Query Service] and xref:services-and-indexes/services/backup-service.adoc[Backup Service]. You can change the availability of memory resources in your cluster based on how you assign your services to each node.

Reviewer: ```suggestion
You must configure your memory quota allocations for each service in Couchbase Server, except for the xref:services-and-indexes/services/query-service.adoc[Query Service] and xref:services-and-indexes/services/backup-service.adoc[Backup Service]. 
You can change the availability of memory resources in your cluster based on how you assign your services to each node.
```

> The memory quota you allocate for a service applies to every instance of that service across your cluster. For example, if you allocate 2048 MB to the xref:services-and-indexes/services/analytics-service.adoc[Analytics Service], and you run the Analytics Service on three of a cluster's five nodes, each instance of the service has 2048 MB of memory.

Reviewer: ```suggestion
The memory quota you allocate for a service applies to every instance of that service across your cluster. 
For example, if you allocate 2048 MB to the xref:services-and-indexes/services/analytics-service.adoc[Analytics Service], and you run the Analytics Service on three of a cluster's five nodes, each instance of the service has 2048 MB of memory.
```

NOTE: You can't allocate different amounts of memory for different instances of the same service in a cluster. 

Couchbase recommends that you allocate no more than 90% of a node's memory (80% on nodes with a small amount of total memory) to a server and its services. 

The firm limits for server memory allocation can be calculated by: stem:[max(total\_memory - 1, 80\% \times total\_memory)] where `total_memory` is the maximum memory on the node in GiB.

> The xref:services-and-indexes/services/data-service.adoc[Data Service] must run on at least one node in any cluster. Every xref:buckets.adoc[bucket] you create on a node has its own memory quota. The available memory for a bucket comes from the quota you assign to the Data Service. For more information on bucket memory quotas, see the <<bucket-memory,Bucket Memory Quotas section>>.

Reviewer: ```suggestion
The xref:services-and-indexes/services/data-service.adoc[Data Service] must run on at least one node in any cluster. Every xref:buckets.adoc[bucket] you create on a node has its own memory quota. 
The available memory for a bucket comes from the quota you assign to the Data Service. 
For more information on bucket memory quotas, see the <<bucket-memory,Bucket Memory Quotas section>>.
```

> When you add a new node, you can choose to use the same configuration from the first node in the cluster. If this default configuration exceeds the memory available for the node, you can customize the new node's settings and change its assigned services. You can add more nodes to a cluster to add more services. 

Reviewer: ```suggestion
When you add a new node, you can choose to use the same configuration from the first node in the cluster. 
If this default configuration exceeds the memory available for the node, you can customize the new node's settings and change its assigned services. 
You can add more nodes to a cluster to add more services. 
```

For more information about how to add nodes and allocate memory to a service when you initialize a cluster, see xref:manage:manage-nodes/create-cluster.adoc[].

> The *Memory Quotas* panel in the xref:manage:manage-settings/general-settings.adoc[General settings screen] of the Couchbase Server Web Console lists all the services running on a cluster. It also shows the memory allocated to each service. You can use the panel to change the memory allocations in your cluster. If a change doesn't meet the required minimum or exceeds the available memory for a cluster, the Memory Quotas panel displays an error. 

Reviewer: ```suggestion
The *Memory Quotas* panel in the xref:manage:manage-settings/general-settings.adoc[General settings screen] of the Couchbase Server Web Console lists all the services running on a cluster. 
It also shows the memory allocated to each service. 
You can use the panel to change the memory allocations in your cluster. 
If a change doesn't meet the required minimum or exceeds the available memory for a cluster, the Memory Quotas panel displays an error. 
```

[#service-memory-quotas]
== Service Memory Quotas

You must configure your memory quota allocations for each service in Couchbase Server, except for the xref:services-and-indexes/services/query-service.adoc[Query Service] and xref:services-and-indexes/services/backup-service.adoc[Backup Service]. You can change the availability of memory resources in your cluster based on how you assign your services to each node.

The memory quota you allocate for a service applies to every instance of that service across your cluster. 
For example, if you allocate 2048 MB to the xref:services-and-indexes/services/analytics-service.adoc[Analytics Service], and you run the Analytics Service on three of a cluster's five nodes, each instance of the service has 2048 MB of memory.

NOTE: You can't allocate different amounts of memory for different instances of the same service in a cluster. 

Couchbase recommends that you allocate no more than 90% of a node's memory (80% on nodes with a small amount of total memory) to a server and its services. 

> The firm limits for server memory allocation can be calculated by: stem:[max(total\_memory - 1, 80\% \times total\_memory)] where `total_memory` is the maximum memory on the node in GiB.

Reviewer 2: The STEM formula breaks the line, leaving an ugly gap at the end of the previous line. Is there any way to shorten the preceding text to avoid the line break? Or put the formula in a line by itself?

Author: I like line by itself.

The xref:services-and-indexes/services/data-service.adoc[Data Service] must run on at least one node in any cluster. Every xref:buckets.adoc[bucket] you create on a node has its own memory quota. 
The available memory for a bucket comes from the quota you assign to the Data Service. 

> For more information on bucket memory quotas, see the <<bucket-memory,Bucket Memory Quotas section>>.

Reviewer 2: Not a big deal, but we don't tend to use the word "section" within the link text. Outside the link text is fine.

When you add a new node, you can choose to use the same configuration from the first node in the cluster. 
If this default configuration exceeds the memory available for the node, you can customize the new node's settings and change its assigned services. 

> You can add more nodes to a cluster to add more services. 

Reviewer 2: This is true, but some administrators deliberately set up different services on different nodes. So when they add a new node, they will as you say "customize the new node's settings and change its assigned services", but it won't be because "the default configuration exceeds the memory available for the node"; it's because they _don't want to use_ the default configuration on the new node. Need to rewrite to reflect this.

NOTE: You can't allocate different amounts of memory for different instances of the same service in a cluster. 

Couchbase recommends that you allocate no more than 90% of a node's memory (80% on nodes with a small amount of total memory) to a server and its services. 

The firm limits for server memory allocation can be calculated by: 

stem:[max(total\_memory - 1, 80\% \times total\_memory)] 

where `total_memory` is the maximum memory on the node in GiB.

The xref:services-and-indexes/services/data-service.adoc[Data Service] must run on at least one node in any cluster. Every xref:buckets.adoc[bucket] you create on a node has its own memory quota. 
The available memory for a bucket comes from the quota you assign to the Data Service. 
For more information on bucket memory quotas, see the <<bucket-memory,Bucket Memory Quotas>> section.

When you add a new node, you can use the same configuration and services from the first node in the cluster. 
You can also choose to customize the new node's settings and change its assigned services. 

> You can add more nodes to a cluster to add more services. 

Reviewer 2: The first sentence sums everything up by itself; don't think we need the second one here.

```suggestion
You can also choose to customize the new node's settings and change its assigned services.
```

| 256
|===

The Query Service and the Backup Service don't require an administrator-specified memory quota.

[#bucket-memory]
== Bucket Memory Quotas 

> You must specify a memory quota for each bucket you create. This quota is allocated for the bucket on a per node basis and must be less than the total memory quota for the cluster. 

Reviewer: ```suggestion
You must specify a memory quota for each bucket you create. 
This quota is allocated for the bucket on a per node basis and must be less than the total memory quota for the cluster. 
```

> Set the memory quota based on the expected size of your dataset. The memory quota for a bucket must support the minimum memory resident ratio of its xref:learn:buckets-memory-and-storage/storage-engines.adoc[storage engine]: 

Reviewer: ```suggestion
Set the memory quota based on the expected size of your dataset. 
The memory quota for a bucket must support the minimum memory resident ratio of its xref:learn:buckets-memory-and-storage/storage-engines.adoc[storage engine]: 
```

* *Couchstore*: The memory quota is recommended to be at least 10-20% of your expected dataset size. 
* *Magma*: The memory quota is recommended to be at least 2% of your expected dataset size. 

For example, if you expect to have about 2TBs of data per node in your cluster and want to use the *Magma* engine, you could set the memory quota for a bucket to 40GB. 

> NOTE: These values are recommendations only. The specific memory quota requirements for your bucket are dependent on access patterns, data density, and other factors.

Reviewer: ```suggestion
NOTE: These values are recommendations only. 
The specific memory quota requirements for your bucket are dependent on access patterns, data density, and other factors.
```

== Ejection

If a bucket's memory quota is exceeded, items may be _ejected_ from the bucket by the Data Service.

> Different ejection methods are available, and are configured per bucket. Note that in some cases, ejection is configured _not_ to occur.

Reviewer: ```suggestion
Different ejection methods are available, and are configured per bucket. 
Note that in some cases, ejection is configured _not_ to occur.
```

For a Couchbase bucket, you can choose betweeen a *Value-only* or *Full* ejection method: 

> * *Value-only*: The bucket only ejects data when it removes a document from memory. Choose this method if you need better performance, but be aware that it uses more system memory. 

Reviewer: ```suggestion
* *Value-only*: The bucket only ejects data when it removes a document from memory. 
Choose this method if you need better performance, but be aware that it uses more system memory. 
```

> * *Full*: The bucket ejects data, metadata, keys, and values when it removes a document from memory. Choose this method if you want to reduce your memory overhead requirement. 

Reviewer: ```suggestion
* *Full*: The bucket ejects data, metadata, keys, and values when it removes a document from memory. 
Choose this method if you want to reduce your memory overhead requirement. 
```
